snuba/clusters/cluster.py

```diff
@@ -374,16 +376,37 @@ def __get_cluster_nodes(self, cluster_name: str) -> Sequence[ClickhouseNode]:
     if (s not in DEV_STORAGE_SETS or settings.ENABLE_DEV_FEATURES)
 }
 
-# Map all storages to clusters via storage sets
-_STORAGE_SET_CLUSTER_MAP = {
-    storage_set: cluster
-    for cluster in CLUSTERS
-    for storage_set in cluster.get_storage_set_keys()
-}
+_STORAGE_SET_CLUSTER_MAP: Optional[Dict[StorageSetKey, ClickhouseCluster]] = None
+
+
+def _get_storage_set_cluster_map() -> Dict[StorageSetKey, ClickhouseCluster]:
+    global _STORAGE_SET_CLUSTER_MAP
+    # Map all storages to clusters via storage sets
+    if _STORAGE_SET_CLUSTER_MAP is None:
+        _STORAGE_SET_CLUSTER_MAP = {
+            storage_set: cluster
+            for cluster in CLUSTERS
+            for storage_set in cluster.get_storage_set_keys()
+        }
+    return _STORAGE_SET_CLUSTER_MAP
```

Comment:
Why is this change needed? Wouldn't it be better to just mark the invalid cluster so that all the other valid clusters are on the map rather than not populate the map at all.

Reply:
Why would the map not be populated at all?

Reply:
This map does get populated except instead of at import time, it's populated at access time.

Reply:
My apologies it is, I assumed you were handling some error and not populating it. So then, why the change at all? Isn't it more straightforward the original way?

Reply:
I think this is an artifact of me trying to get the whole thing working end to end. You're probably right that this change is probably unnecessary and I can remove it if you feel strongly but here's why I think it should stay:

1. I try to minimize logic at import time unless it's absolutely necessary
2. _STORAGE_SET_CLUSTER_MAP is a global variable and now it's explicitly labeled as such
3. having this be a function makes it easier to mock for tests (In fact I've already used it in one of my commits addressing your comments)

But again if you say you'd still like it removed I can remove it.

Reply:
I'm not sure I agree with those points. For example, we do 1 extensively in the Snuba codebase (all storages, entities and datasets). Unclear to me why it's a problem here. Since it's not relevant to this change I think it's better to remove it.

Reply:
Well for point number 3 I already used it to help me write a test for the improperly caught exception you pointed out

https://github.com/getsentry/snuba/pull/2772/files#diff-3dcdd4ef5de62043035d4ab2e57d185826e6c7d147b21cb7a4824f7e956f2666R12

Reply:
I put the map creation logic back to the import level but I still kept its access behind a function because it makes testing easier